doc/edgehub-api.md

```diff
@@ -0,0 +1,434 @@
+
+Let's say that a device has just connected to Iot Hub/Edge Hub and it pulled the entire twin. It receives something like the following:
+
+```
+"properties": {
+    "desired": {
+        "temperature": 5,
+        $version: 23
+    },
+    "reported": {
+        "temperature": 7
+        $version: 98
+    }
+}
+```
+
+Compared to the earlier example, this desired properties section has a $version field with a numeric value. The above mentioned patch also always have a version number included:
+
+```
+{ "some-property": 123, "$version" : 25 }
+```
+
+Comparing the result of the twin pull operation and the desired property patch, one has the version 23 and the other 25. That means that between the twin pull and the patch there might have been another change.
+
+To avoid missing patches, the recommended way of using twins is the following:
+- Subscribe for desired property changes
+- Pull the entire twin, parse the desired property changes section and store the $version
+- Going forward check the $version propery of every incoming patch. If the version number incremented by more then one, there are odds that a patch was missed, so pull the entire twin.
+
+## Cloud-to-Device messages
+Twin is a simple way to send configuration to a device or module, but it is not feasible to send complex information or data with bigger size. If that is a requirement, [cloud-to-device](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-c2d) or C2D messages can be used to send arbitrary data, however these can be sent only to devices, not modules.
+
+C2D messages are much like telemetry messages. A back-end application pushes a message to IoT Hub. IoT Hub enqueues that message and stores it (up to 48 hours). If the device is online, Iot Hub sends the message to the device and waits for a feedback. 
+
+This feedback can be 'completing' the message, 'rejecting' or 'Abandoning' it. Compliting the message means that it was processed succesfully and no need for the message anymore, so IoT Hub deletes it. Rejecting the message means that the device does not want to process the message. In this case IoT Hub moves it in a dead letter queue where it is available for a while for a back end application. Abandoning means that the message cannot be processed right now. In this case IoT Hub puts it back the the queue and tries again later.
+
+If Edge Hub is between the device and Iot Hub, Edge Hub relays the messages and the feedbacks. It does not 'download' the C2D messages so it can forward them later. When the device signals Edge Hub that it is ready to receive C2D messages, Edge Hub signals in the name of the device that it is ready to receive C2D messages. When IoT Hub sends a C2D message, then Edge Hub  it to the device and waits for the feedback. When the feedback arrives, Edge Hub propagates it back to IoT Hub.
```

Comment:
```suggestion
If Edge Hub is between the device and Iot Hub, Edge Hub relays the messages and the feedbacks. It does not 'download' the C2D messages so it can forward them later. When the device signals Edge Hub that it is ready to receive C2D messages, Edge Hub signals in the name of the device that it is ready to receive C2D messages. When IoT Hub sends a C2D message, then Edge Hub sends it to the device and waits for the feedback. When the feedback arrives, Edge Hub propagates it back to IoT Hub.
```